### ISSUE TITLE
python crash when funtion from opti (sqpmethod) is evaluated in parallel

### ISSUE THREAD

**Issue by sandeep026** [casadi/casadi#3609]
### Discussed in https://github.com/casadi/casadi/discussions/3608

<div type='discussions-op-text'>

<sup>Originally posted by **sandeep026** March  8, 2024</sup>
Function obtained from opti using to_function() is required to be evaluated in parallel. Used sqpmethod with qpoases, osqp and qrqp.  No issues with serial or unroll or thread with max thread=1.  Could it be a bug?

1. sometimes crashes when 'thread' is used with osqp,qpoaes, highs
2.  no crashes with qrqp, proxqp

A MWE is given below. I am not able to follow the error thrown by the function.

```
import casadi as cs
nlp=cs.Opti()
x=nlp.variable(1)
nlp.minimize(cs.sumsqr(x-5))
nlp.solver('sqpmethod',{'qpsol':'qpoases'})
nlp.solve()
f=nlp.to_function('f',[x],[x,nlp.f])
F=f.map(10,'thread',12)
X=cs.linspace(1,10,10)
Y=F(X)
print(Y)
```

```
qpOASES -- An Implementation of the Online Active Set Strategy.     
Copyright (C) 2007-2015 by Hans Joachim Ferreau, Andreas Potschka,  
Christian Kirches et al. All rights reserved.

qpOASES is distributed under the terms of the
GNU Lesser General Public License 2.1 in the hope that it will be   
useful, but WITHOUT ANY WARRANTY; without even the implied warranty 
of MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.
See the GNU Lesser General Public License for more details.

-------------------------------------------
This is casadi::Sqpmethod.
Using exact Hessian
Number of variables:                               1
Number of constraints:                             0
Number of nonzeros in constraint Jacobian:         0
Number of nonzeros in Lagrangian Hessian:          1

iter      objective    inf_pr    inf_du     ||d||  lg(rg) ls    info
   0   2.500000e+01  0.00e+00  1.00e+01  0.00e+00       -  0  -
WARNING:   Stepsize is 0.000000000000000e+00


#################   qpOASES  --  QP NO.   1   ##################

    Iter   |    StepLength    |       Info       |   nFX
 ----------+------------------+------------------+---------
       0   |   0.000000e+00   |   REM BND    0   |     0
       1   |   1.000000e+00   |    QP SOLVED     |     0
   1   7.888609e-31  0.00e+00  1.78e-15  5.00e+00       -  1  -
MESSAGE(sqpmethod): Convergence achieved after 1 iterations
      solver  :   t_proc      (avg)   t_wall      (avg)    n_eval
          QP  |   1.00ms (  1.00ms) 911.00us (911.00us)         1
  linesearch  |        0 (       0)   7.00us (  7.00us)         1
      nlp_fg  |        0 (       0)   4.00us (  4.00us)         1
  nlp_hess_l  |        0 (       0)   4.00us (  4.00us)         1
  nlp_jac_fg  |        0 (       0)  12.00us (  6.00us)         2
       total  |   1.00ms (  1.00ms)   1.44ms (  1.44ms)         1

qpOASES -- An Implementation of the Online Active Set Strategy.
Copyright (C) 2007-2015 by Hans Joachim Ferreau, Andreas Potschka,
Christian Kirches et al. All rights reserved.

qpOASES is distributed under the terms of the
GNU Lesser General Public License 2.1 in the hope that it will be
useful, but WITHOUT ANY WARRANTY; without even the implied warranty
of MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.
See the GNU Lesser General Public License for more details.

-------------------------------------------
This is casadi::Sqpmethod.
Using exact Hessian
Number of variables:                               1
Number of constraints:                             0
Number of nonzeros in constraint Jacobian:         0
Number of nonzeros in Lagrangian Hessian:          1

iter      objective    inf_pr    inf_du     ||d||  lg(rg) ls    info
iter      objective    inf_pr    inf_du     ||d||  lg(rg) ls    info
iter      objective    inf_pr    inf_du     ||d||  lg(rg) ls    info
iter      objective    inf_pr    inf_du     ||d||  lg(rg) ls    info
   0   1.600000e+01  0.00e+00  8.00e+00  0.00e+00       -  0  -
iter      objective    inf_pr    inf_du     ||d||  lg(rg) ls    info
iter      objective    inf_pr    inf_du     ||d||  lg(rg) ls    info
   0   1.000000e+00  0.00e+00  2.00e+00  0.00e+00       -  0  -
   0   4.000000e+00  0.00e+00  4.00e+00  0.00e+00       -  0  -
iter      objective    inf_pr    inf_du     ||d||  lg(rg) ls    info
   0   9.000000e+00  0.00e+00  6.00e+00  0.00e+00       -  0  -
iter      objective    inf_pr    inf_du     ||d||  lg(rg) ls    info
iter      objective    inf_pr    inf_du     ||d||  lg(rg) ls    info
WARNING:  QProblem has already been initialised
   0   1.000000e+00  0.00e+00  2.00e+00  0.00e+00       -  0  -
WARNING:  QProblem has already been initialised
   0   2.500000e+01  0.00e+00  1.00e+01  0.00e+00       -  0  -
WARNING:  QProblem has already been initialised
   0   1.600000e+01  0.00e+00  8.00e+00  0.00e+00 WARNING:  QProblem has already been initialised
WARNING:   Stepsize is 0.000000000000000e+00


#################   qpOASES  --  QP NO.   5   ##################

    Iter   |    StepLength    |       Info       |   nFX
 ----------+------------------+------------------+---------
WARNING:  QProblem has already been initialised
WARNING:   Stepsize is 0.000000000000000e+00
   0   9.000000e+00  0.00e+00  6.00e+00  0.00e+00 WARNING:   Stepsize is 0.000000000000000e+00
ERROR:  The error occurred is not yet known
->ERROR:  Removing from active set failed
WARNING:   Stepsize is 0.000000000000000e+00

       0   |   0.000000e+00   |   REM BND    0   |     0
WARNING:   Stepsize is 0.000000000000000e+00
ERROR:  The error occurred is not yet known
      -       -  0  -
ERROR:  The error occurred is not yet known
  ->ERROR:  Removing from active set failed
->ERROR:  Removing from active set failed
WARNING:   Stepsize is 0.000000000000000e+00

ERROR:  The error occurred is not yet known
->ERROR:  Removing from active set failed
       1   |   1.000000e+00   |    QP SOLVED     |     0


#################   qpOASES  --  QP NO.   6   ##################

    Iter   |    StepLength    |       Info       |   nFX
 ----------+------------------+------------------+---------
      ->ERROR:  Unable to perform homotopy step
  ->ERROR:  Unable to perform homotopy step
->WARNING:  QProblem has already been initialised

WARNING:   Stepsize is 0.000000000000000e+00
 0       0   |   1.000000e+00   |   REM BND    0   |     0
   0   0.000000e+00  0.00e+00  0.00e+00  0.00e+00       -  0  -
    ->ERROR:  Unable to perform homotopy step
        1   |   1.000000e+00   |    QP SOLVED     |     0
    ->ERROR:  Removing from active set failed
  ->ERROR:  Initialisation failed! QP could not be solved!
   1   0.000000e+00  0.00e+00  0.00e+00  3.00e+00 MESSAGE(sqpmethod): Convergence achieved after 0 iterations


#################   qpOASES  --  QP NO.   7   ##################

    Iter   |    StepLength    |       Info       |   nFX
 ----------+------------------+------------------+---------
   1   0.000000e+00  0.00e+00  0.00e+00  4.00e+00 iter      objective    inf_pr    inf_du     ||d||  lg(rg) ls    info
      ->ERROR:  Removing from active set failed
      ->ERROR:  Initialisation failed! QP could not be solved!
      solver  :   t_proc      (avg)   t_wall      (avg)    n_eval
  nlp_jac_fg        0   |   1.000000e+00   |   REM BND    0   |     0
      -  1  -
MESSAGE(sqpmethod): Convergence achieved after 1 iterations
->ERROR:  Removing from active set failed
          ->ERROR:  Initialisation failed! QP could not be solved!
   0   4.000000e+00  0.00e+00  4.00e+00  0.00e+00       -  0  -
 -
Function qpsol (0x1e0c18fe440)
Input 0 (h):       - Function qpsol (0x1e0c18fe440)
Input 0 (h):       solver             -> |        0 (       0)   9.00us (  9.00us)         1
       total             ->WARNING:   Stepsize is 0.000000000000000e+00

 1  -
WARNING:   Stepsize is 0.000000000000000e+00
ERROR:  The error occurred is not yet known
->ERROR:  Removing from active set failed
2
Input 1 (g):  |  11.00ms ( 11.00ms)  11.24ms ( 11.24ms)         1
  ->ERROR:  Unable to perform homotopy step


#################   qpOASES  --  QP NO.   9   ##################

    Iter   |    StepLength    |       Info       |   nFX
 ----------+------------------+------------------+---------
 :   t_proc      (avg)   t_wall      (avg)    n_eval
MESSAGE(sqpmethod): Convergence achieved after 1 iterations
       1   |   1.000000e+00   |    QP SOLVED     |     0
-4
Input 2 (a): 0x1
Input 3 (lba): 0x1
Input 4 (uba): 0x1
Input 5 (lbx):    1   1.600000e+01  0.00e+00  8.00e+00  0.00e+00   ->    ->ERROR:  Removing from active set failed
      ->ERROR:  Initialisation failed! QP could not be solved!
          QP       solver  :   t_proc      (avg)   t_wall      (avg)    n_eval
          QP ERROR:  Unable to perform homotopy step
          ->ERROR:  Removing from active set failed
            ->ERROR:  Initialisation failed! QP could not be solved!
 |  10.00ms ( 10.00ms)   9.97ms (  9.97ms)         1
  linesearch  |        0 (       0)  11.00us ( 11.00us)         1
 |   6.00ms (  6.00ms)   6.12ms (  6.12ms)         1
  linesearch 2
Input 1 (g): Function qpsol (0x1e0c18fe440)
Input 0 (h): Function qpsol (0x1e0c18fe440)
Input 0 (h):       -  1  -
MESSAGE(sqpmethod): Search direction becomes too small without convergence criteria being met.
Function qpsol (0x1e0c18fe440)
Input 0 (h):       nlp_fg -2
Input 2 (a): 0x1
Input 3 (lba): 0x1
Input 4 (uba): 0x1
Input 5 (lbx):        0   |   0.000000e+00   |   REM BND    0   |     0
2
Input 1 (g): -inf
Input 6 (ubx):  |        0 (       0)   9.00us (  9.00us)         1
10
Input 2 (a): 0x1
Input 3 (lba): 0x1
Input 4 (uba): 0x1
Input 5 (lbx):       solver -inf
Input 6 (ubx): -inf
Input 6 (ubx):        1   |   1.000000e+00   |    QP SOLVED     |     0
inf
Input 7 (x0): inf
Input 7 (x0):       nlp_fg -4
Input 8 (lam_x0): 4
Input 8 (lam_x0): inf
Input 7 (x0):  |        0 (       0)   6.00us (  6.00us)         1
-0
Input 9 (lam_a0): 0x1
Input 10 (q): NULL
Input 11 (p): NULL
-0
Input 9 (lam_a0): 0x1
Input 10 (q): NULL
Input 11 (p): NULL
Function solver (0x1e0c18fa1b0)
Input 0 (x0): 2
Input 1 (g):  :   t_proc      (avg)   t_wall      (avg)    n_eval
10
Input 1 (p): 0x1
Input  |        0 (       0)   6.00us (  6.00us)         1
Function solver (0x1e0c18fa1b0)
Input 0 (x0):           QP 2
Input 1 (g): 3
Input 1 (p): 0x1
Input 2 (lbx): 4
Input 2 (a): 0x1
Input 3 (lba): 0x1
Input 4 (uba): 0x1
Input 5 (lbx): 4
Input 8 (lam_x0):  |  12.00ms ( 12.00ms)  11.75ms ( 11.75ms)         1
  linesearch   nlp_hess_l 2 (lbx): -inf
Input 3 (ubx): 2
Input 2 (a): 0x1
Input 3 (lba): 0x1
Input 4 (uba): 0x1
Input 5 (lbx): -0
Input 9 (lam_a0): 0x1
Input 10 (q): NULL
Input 11 (p): NULL
-inf
Input 6 (ubx): -inf
Input 6 (ubx): -inf
Input 3 (ubx): inf
Input 4 (lbg): 0x1
Input 5 (ubg): 0x1
Input 6 (lam_x0): inf
Input 7 (x0): inf
Input 4 (lbg): 0x1
Input 5 (ubg): 0x1
Input 6 (lam_x0): 0
Input 7 (lam_g0): 0x1
 |        0 (       0)  14.00us ( 14.00us)         1
 |        0 (       0)   3.00us (  3.00us)         1
  nlp_jac_fg    1   0.000000e+00  0.00e+00  0.00e+00  3.00e+00       -  1  -
      nlp_fg   nlp_hess_l  |        0 (       0)  10.00us (  5.00us)         2
       total 0
Input 7 (lam_g0): 0x1
Function f (0x1e0c18ff5d0)
Input 0 (i0):  |        0 (       0)   1.00us (  1.00us)         1
  nlp_jac_fg  |  14.00ms ( 14.00ms)  13.59ms ( 13.59ms)         1
inf
Input 7 (x0):  |        0 (       0)   8.00us (  4.00us)         2
       total Function f (0x1e0c18ff5d0)
Input 0 (i0): 10
0
Input 8 (lam_x0): CasADi - 0
Input 8 (lam_x0): 2024- |  19.00ms ( 19.00ms)  18.79ms ( 18.79ms)         1
MESSAGE(sqpmethod): Convergence achieved after 1 iterations
 |        0 (       0)   8.00us (  8.00us)         1
3
-0
Input 9 (lam_a0): 0x1
Input 10 (q): NULL
Input 11 (p): NULL
Function solver (0x1e0c18fa1b0)
Input 0 (x0): CasADi - Function solver (0x1e0c18fa1b0)
Input 0 (x0): 4
Input 1 (p): 0x1
Input 2 (lbx):   nlp_hess_l -0
Input 9 (lam_a0): 0x1
Input 10 (q): NULL
Input 11 (p): NULL
7
Input 1 (p): 0x1
Input 2 (lbx): -inf
Input 3 (ubx):  |        0 (       0)   6.00us (  6.00us)         1
2024-03-08 15:20:16 WARNING("  nlp_jac_fg Function solver (0x1e0c18fa1b0)
Input 0 (x0): Exception raised: Error in Function::operator() for 'f' [MXFunction] at .../casadi/core/function.cpp:1482:
Error in Function::operator() for 'solver' [Sqpmethod] at .../casadi/core/function.cpp:1482:
Error in Function::operator() for 'qpsol' [QpoasesInterface] at .../casadi/core/function.cpp:1482:
.../casadi/core/conic.cpp:537: conic process failed. Set 'error_on_fail' option to false to ignore this error.") [.../casadi/core/map.cpp:453]
-inf
Input 3 (ubx): 6
Input 1 (p): 0x1
Input 2 (lbx):  |        0 (       0)   4.00us (  2.00us)         2
03-08 15:20:16 WARNING("       total       solver inf
Input 4 (lbg): 0x1
Input 5 (ubg): 0x1
Input 6 (lam_x0):  :   t_proc      (avg)   t_wall      (avg)    n_eval
          QP -inf
Input 3 (ubx):  |  24.00ms ( 24.00ms)  24.24ms ( 24.24ms)         1
0
Input 7 (lam_g0): 0x1
Exception raised: Error in Function::operator() for 'f' [MXFunction] at .../casadi/core/function.cpp:1482:
Error in Function::operator() for 'solver' [Sqpmethod] at .../casadi/core/function.cpp:1482:
Error in Function::operator() for 'qpsol' [QpoasesInterface] at .../casadi/core/function.cpp:1482:
.../casadi/core/conic.cpp:537: conic process failed. Set 'error_on_fail' option to false to ignore this error.") [.../casadi/core/map.cpp:453]
Function f (0x1e0c18ff5d0)
Input 0 (i0): inf
Input 4 (lbg):  |  16.00ms ( 16.00ms)  16.59ms ( 16.59ms)         1
inf
Input 4 (lbg): 0x1
Input 5 (ubg): 0x1
Input 6 (lam_x0): 4
  linesearch CasADi - 0x1
Input 5 (ubg): 0x1
Input 6 (lam_x0): 2024-03-08 15:20:17 WARNING("0
Input 7 (lam_g0): 0x1
0Function f (0x1e0c18ff5d0)
Input 0 (i0):
Input 7 (lam_g0): 0x1
Exception raised: Error in Function::operator() for 'f' [MXFunction] at .../casadi/core/function.cpp:1482:
Error in Function::operator() for 'solver' [Sqpmethod] at .../casadi/core/function.cpp:1482:
Error in Function::operator() for 'qpsol' [QpoasesInterface] at .../casadi/core/function.cpp:1482:
.../casadi/core/conic.cpp:537: conic process failed. Set 'error_on_fail' option to false to ignore this error.") [.../casadi/core/map.cpp:453]
7
Function f (0x1e0c18ff5d0)
Input 0 (i0):  |        0 (       0)   9.00us (  9.00us)         1
CasADi - 6
2024-03-08 15:20:17 WARNING("CasADi - 2024-03-08 15:20:17 WARNING("      nlp_fg Exception raised: Error in Function::operator() for 'f' [MXFunction] at .../casadi/core/function.cpp:1482:
Error in Function::operator() for 'solver' [Sqpmethod] at .../casadi/core/function.cpp:1482:
Error in Function::operator() for 'qpsol' [QpoasesInterface] at .../casadi/core/function.cpp:1482:
.../casadi/core/conic.cpp:537: conic process failed. Set 'error_on_fail' option to false to ignore this error.") [.../casadi/core/map.cpp:453]
 |        0 (       0)   6.00us (  6.00us)         1
Exception raised: Error in Function::operator() for 'f' [MXFunction] at .../casadi/core/function.cpp:1482:
Error in Function::operator() for 'solver' [Sqpmethod] at .../casadi/core/function.cpp:1482:
Error in Function::operator() for 'qpsol' [QpoasesInterface] at .../casadi/core/function.cpp:1482:
.../casadi/core/conic.cpp:537: conic process failed. Set 'error_on_fail' option to false to ignore this error.") [.../casadi/core/map.cpp:453]
  nlp_hess_l  |   1.00ms (  1.00ms)   6.00us (  6.00us)         1
  nlp_jac_fg  |        0 (       0)   3.00us (  1.50us)         2
       total  |  56.00ms ( 56.00ms)  55.55ms ( 55.55ms)         1
Traceback (most recent call last):
  File "c:\Users\Sandeep\Documents\NonlinearOptimization\ocp2nlp\docs\examples\analytical_solution\detest.py", line 11, in <module>    
    Y=F(X)
      ^^^^
  File "c:\Users\Sandeep\Documents\NonlinearOptimization\.venv\Lib\site-packages\casadi\casadi.py", line 23381, in __call__
    ret = self.call(args)
          ^^^^^^^^^^^^^^^
  File "c:\Users\Sandeep\Documents\NonlinearOptimization\.venv\Lib\site-packages\casadi\casadi.py", line 20039, in call
    return _casadi.Function_call(self, *args)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
RuntimeError: Error in Function::call for 'threadmap10_f' [ThreadMap] at .../casadi/core/function.cpp:330:
.../casadi/core/function_internal.hpp:1616: Evaluation failed
```</div>

**Comment by jgillis** [comment]
A regression bug related to this was fixed in 3.6.5. Coule you validate if this issue still exists?

**Comment by sandeep026** [comment]
Tested 3.6.5 and this issue has been fixed and I can confirm it. Works as expected :thumbsup:
